Nr 366597 automated nrjmx releases

rahulreddy15 opened the pull request:
Automating the release of `nrjmx`.

There were issues with using the [reusable_on_release.yaml](https://github.com/newrelic/coreint-automation/blob/main/.github/workflows/reusable_on_release.yaml) and [reusable_pre_release.yaml](https://github.com/newrelic/coreint-automation/blob/main/.github/workflows/reusable_pre_release.yaml) from `coreint-automation`

Reusable Pre Release could not be used because
- `test-nix` job which runs unit test on linux does not setup docker credentials which are required for `nrjmx`
- `test-windows` and `test-integration-nix` would be skipped because nrjmx runs these via docker images
- `build-win-packages` could not be reused because the make command as well as the scripts that upload assets to GH work differently
- `publish` could not be reused because of the way appname is set. The prefix `nri-` is always added

Reusable Release could not be reused because
- `publish` could not be reused because of the way appname is set. The prefix `nri-` is always added. 

Modifying these action for the special case of `nrjmx` does not make sense.
Instead of using the reusable workflows this PR focuses on automating the release part.

The automated release action is used to schedule the releases.
The steps like updating the GH prerelease title on successful prerelease, notifying in the case of failed pre release / release, and notifying success in the case of successful release are added.
Fixed the `CHANGELOG` to be acceptable to the automated release action by adding an `Unreleased` section and modifying the format of older entries.

Also Included in the PR:
Updated all the github action versions
Dropped the use of snyk in security testing and moved to trivy like the rest of the integrations.

sairaj18 commented on .github/workflows/prerelease.yml:
```diff
@@ -169,3 +169,34 @@ jobs:
           # used for signing package stuff
           gpg_passphrase: ${{ env.GPG_PASSPHRASE }}
           gpg_private_key_base64: ${{ env.GPG_PRIVATE_KEY_BASE64 }}
+          
```
In coreint-automation [Test package from Staging repo](https://github.com/newrelic/coreint-automation/blob/ec8b71baa82d846f6fd9cf0262792070ec0bf3c7/.github/workflows/reusable_pre_release.yaml#L264C15-L264C45) job is ran before updating title of successful pre-release. Don't we need it here?

rahulreddy15 replied:
Added the package installation testing step.

sairaj18 commented on .github/workflows/prerelease.yml:
```diff
@@ -169,3 +169,34 @@ jobs:
+
+  notify-failure:
+    if: ${{ always() && failure() }}
+    needs: [unit-test-linux, test-go-linux-jdk8, test-go-linux-jdk11, check-gen-code, publishing-to-s3]
```
Why are `packaging` and `packaging-msi` jobs not included here

rahulreddy15 replied:
I did not add those because `publishing-to-s3` would fail if one of those failed.
But you're right. Better to include them.

sairaj18 commented on .github/workflows/prerelease.yml:
```diff
@@ -169,3 +169,34 @@ jobs:
+
+  update-title-on-failure:
+    if: ${{ always() && failure() }}
+    needs: [unit-test-linux, test-go-linux-jdk8, test-go-linux-jdk11, check-gen-code, publishing-to-s3]
```
Why are `packaging` and `packaging-msi` jobs not included here

rahulreddy15 replied:
I did not add those because `publishing-to-s3` would fail if one of those failed.
But you're right. Better to include them.

sairaj18 commented on .github/workflows/release.yml:
```diff
@@ -58,3 +58,22 @@ jobs:
           # used for signing package stuff
           gpg_passphrase: ${{ env.GPG_PASSPHRASE }}
           gpg_private_key_base64: ${{ env.GPG_PRIVATE_KEY_BASE64 }}
+
```
In coreint-automation [Test package from prod repo](https://github.com/newrelic/coreint-automation/blob/ec8b71baa82d846f6fd9cf0262792070ec0bf3c7/.github/workflows/reusable_on_release.yaml#L126) job is ran before updating title of successful pre-release. Don't we need it here?

rahulreddy15 replied:
Added the package installation testing step.

sairaj18 commented on .github/workflows/security.yml:
```diff
@@ -16,19 +11,41 @@ on:
+          severity: 'HIGH,CRITICAL'
+        env:
+          TRIVY_DB_REPOSITORY: public.ecr.aws/aquasecurity/trivy-db
+          TRIVY_JAVA_DB_REPOSITORY: public.ecr.aws/aquasecurity/trivy-java-db
```
The coreint-automation has [Upload Trivy scan results to GitHub Security tab](https://github.com/newrelic/coreint-automation/blob/ec8b71baa82d846f6fd9cf0262792070ec0bf3c7/.github/workflows/reusable_security.yaml#L57C15-L57C63) job in the reuseable_security workflow. Is is not needed here?

rahulreddy15 replied:
We only need to upload the results that way if we are running the workflow periodically ( on a schedule ).
The pre-release and release workflows are not run on a schedule.

sairaj18 replied:
> We only need to upload the results that way if we are running the workflow periodically ( on a schedule ).

Do you know what is the use of uploading those results?

If the job fails when it detects the vulnerability it should be fine even don't upload the results. I'm not sure what is the use from uploading the trivy results. So, I'm not sure if we should upload or not Just added comment because it was there in coreint-automation

sairaj18 commented on CHANGELOG.md:
```diff
@@ -1,66 +1,80 @@
+## Unreleased
+
+### enhancements
+- Moved to automated releases for nrjmx
```
This should not be considered as an enhancement as the workflow file changes are not the actual code changes i.e either dependency bump or new feature added or fixing bug.

For release toolkit to detect changelog and create new version pre-release and release U can bump any dependency and added it here.

rahulreddy15 replied:
It is an enhancement to the codebase.
Including that here is fine and have done it before with `nri-statsd` so that the automatic release is created.

sairaj18 replied:
I also added it for [nri-jmx](https://github.com/newrelic/nri-jmx/pull/158#discussion_r1853389752). But christian asked me to remove it because it is not what customers use right? i.e the binary doesn't have any changes even if we are enhancing the codebase

rahulreddy15 replied:
Made this change

sairaj18 commented:
In pre-release.yml

```
on:
  release:
    types:
      - prereleased
    tags:
      - '*'
```

In other integrations the tags is not * instead it is `v*`

```
on:
  release:
    types:
      - prereleased
    tags:
      - 'v*'
```

sairaj18 commented:
In security.yml

```
on:
  push:
    branches:
      - '**'
    tags-ignore:
      - '**'
    paths-ignore:
      - README.md
```

In other integrations for example [nri-redis](https://github.com/newrelic/nri-redis/blob/master/.github/workflows/security.yaml) has the trigger different from the one nrjmx has. Do u think we need to update it?

```
on:
  push:
    branches:
      - master
      - main
      - renovate/**
  pull_request:
  schedule:
    - cron: "0 3 * * *"
```

rahulreddy15 commented:
> push:
>     branches:
>       - master
>       - main
>       - renovate/**
>   pull_request:

Made this change.

rahulreddy15 commented:
> In security.yml
> 
> ```
> on:
>   push:
>     branches:
>       - '**'
>     tags-ignore:
>       - '**'
>     paths-ignore:
>       - README.md
> ```
> 
> In other integrations for example [nri-redis](https://github.com/newrelic/nri-redis/blob/master/.github/workflows/security.yaml) has the trigger different from the one nrjmx has. Do u think we need to update it?
> 
> ```
> on:
>   push:
>     branches:
>       - master
>       - main
>       - renovate/**
>   pull_request:
>   schedule:
>     - cron: "0 3 * * *"
> ```

We are not scheduling the security check so:
1. We don't need to upload the results, that is automatic.
2. Removed tags ignore and paths ignore because they are not relevant.
3. Added pull_request because that makes sense here

sairaj18 commented on .github/workflows/security.yml:
```diff
@@ -1,34 +1,48 @@
 on:
   push:
     branches:
       - '**'
```
Can this be triggered for only master, main, renovate/** refer [nri-redis](https://github.com/newrelic/nri-redis/blob/master/.github/workflows/security.yaml)

rahulreddy15 replied:
Changed this to be similar to `nri-redis`

sairaj18 reviewed (approved):
LGTM 🚀

sairaj18 commented:
One thing that i've noticed is that under releases we r having `noarch` in the package names. I think we might need to look into why it is not able to see arch in package names
![image](https://github.com/user-attachments/assets/45257e06-47eb-4589-b13f-88e1748c366e)

davidgit commented:
This PR was merged without waiting for a test that had passed for the previous commit.

Approved

rajrohanyadav commented:
approved.
